ssl/s3_enc.c

```diff
@@ -399,11 +418,14 @@ int ssl3_change_cipher_state(SSL *s, int which)
 
     OPENSSL_cleanse(&(exp_key[0]), sizeof(exp_key));
     OPENSSL_cleanse(&(exp_iv[0]), sizeof(exp_iv));
-    EVP_MD_CTX_cleanup(&md);
     return (1);
  err:
     SSLerr(SSL_F_SSL3_CHANGE_CIPHER_STATE, ERR_R_MALLOC_FAILURE);
  err2:
+    if (s->enc_write_ctx && !(which & SSL3_CC_READ)) {
+        EVP_CIPHER_CTX_free(s->enc_write_ctx);
+        s->enc_write_ctx = NULL;
+    }
```

Comment:
If I change the UA Stack to check if SSL_read or SSL_write returns SSL_ERROR_SSL/SYSCALL
and avoid SSL_shutdown in that case, the crash is no longer reproducible.
So I am fine with reverting this hunk.